OMCompiler/Compiler/NBackEnd/NBEquation.mo

```diff
@@ -207,78 +206,174 @@ public
 
         case qualEq as IF_EQUATION()
           algorithm
-            qualEq.conditions := List.map(qualEq.conditions, funcExp);
-            for eqn_lst in qualEq.eqnstrue loop
-              eqnstrue := List.map2(eqn_lst, map, funcExp, funcCrefOpt) :: eqnstrue;
-            end for;
-            qualEq.eqnstrue := listReverse(eqnstrue);
-            qualEq.eqnsfalse := List.map2(qualEq.eqnsfalse, map, funcExp, funcCrefOpt);
+            ifEqBody := IfEquationBody.map(qualEq.body, funcExp, funcCrefOpt);
+            if not referenceEq(ifEqBody, qualEq.body) then
+              qualEq.body := ifEqBody;
+            end if;
         then qualEq;
 
         case qualEq as FOR_EQUATION()
           algorithm
-            qualEq.range := funcExp(qualEq.range);
-            qualEq.body := map(qualEq.body, funcExp, funcCrefOpt);
+            range := Expression.map(qualEq.range, funcExp);
+            body := map(qualEq.body, funcExp, funcCrefOpt);
+            if not referenceEq(range, qualEq.range) then
+              qualEq.range := range;
+            end if;
+            if not referenceEq(body, qualEq.body) then
+              qualEq.body := body;
+            end if;
         then qualEq;
 
         case qualEq as WHEN_EQUATION()
           algorithm
-            qualEq.body := WhenEquationBody.map(qualEq.body, funcExp, funcCrefOpt);
+            whenEqBody := WhenEquationBody.map(qualEq.body, funcExp, funcCrefOpt);
+            if not referenceEq(whenEqBody, qualEq.body) then
+              qualEq.body := whenEqBody;
+            end if;
         then qualEq;
 
         case qualEq as AUX_EQUATION(body = SOME(body))
           algorithm
-            qualEq.body := SOME(map(body, funcExp, funcCrefOpt));
+            new_body := map(body, funcExp, funcCrefOpt);
+            if not referenceEq(new_body, body) then
+              qualEq.body := SOME(new_body);
+            end if;
         then qualEq;
       end match;
     end map;
-
   end Equation;
 
+  uniontype IfEquationBody
+    record IF_EQUATION_BODY
+      Expression condition                  "the if-condition" ;
+      list<Equation> then_eqns              "body equations";
+      Option<IfEquationBody> else_if        "optional elseif equation";
+    end IF_EQUATION_BODY;
+
+    function toString
+      input IfEquationBody body;
+      input String indent = "";
+      input String elseStr = "";
+      input Boolean selfCall = false;
+      output String str;
+    protected
+      IfEquationBody elseIf;
+    algorithm
+      if not Expression.isEmpty(body.condition) then
+        str := elseStr + "if " + Expression.toString(body.condition) + " then \n";
+      else
+        str := elseStr + "\n";
+      end if;
+      for eqn in body.then_eqns loop
+        str := str + Equation.toString(eqn, indent + "  ") + "\n";
+      end for;
+      if isSome(body.else_if) then
+        SOME(elseIf) := body.else_if;
+        str := str + toString(elseIf, indent, indent +"else ", true);
```

Comment:
This should handle indentations better. Do you agree?
And I believe we can avoid trailing white spaces :)
```suggestion
      input Boolean selfCall = false;
      output String str;
    protected
      IfEquationBody elseIf;
    algorithm
      str := match (Expression.isEmpty(body.condition), selfCall)
        case (false, false)
          then indent + "if " + Expression.toString(body.condition) + " then\n";
        case (false, true)
          then indent + "else if " + Expression.toString(body.condition) + " then\n";
        case (true, true)
          then indent + "else\n";
      end match;
      for eqn in body.then_eqns loop
        str := str + Equation.toString(eqn, indent + "  ") + "\n";
      end for;
      if isSome(body.else_if) then
        SOME(elseIf) := body.else_if;
        str := str + toString(elseIf, indent, true);
```

Reply:
I actually just realized that the new frontend does not have an EMPTY() expression for the last condition but instead just a boolean that is true. Unfortunately i do not want to check for that since it may be that other conditions but the last can be simplified to true...

In general your structure is better i think but i have to adapt this part anyways and don't know if it is feasible for the new structure. :D

Reply:
If the ifs are in the correct order, everything after an `else if true` can be chopped off. Is that what you are getting at?

Not sure what you need to adapt :D
I agree that treating if equations like this is probably going to be more elegant.

Reply:
Well now that i think of it .... it should be like that yes. Originally i thought we need to preserve the rest afterwards, but i think we can just discard everything after a true condition.

Thanks! :D

Reply:
Unfortunately i can't use this since i also use the elseStr input as indicator for the first time it is called. So only the first element has the equation identifier:

```
(1)     [-IF-] if time > 0.5 then
        [----]   [SCAL] x = sin(time)
        [----] else
        [----]   [SCAL] x = 3.0
        [----] end if;
```

Reply:
But i updated the check for `true` and added another for `false`, most of this is done in the frontend, but it is not expensive so i might as well just check for it.

Reply:
Oh right, I didn't see that. That's a nice way to print equations.

Finding those `true` and `false` branches might even be useful information for the modeller. Should we give a warning that certain branches are never reached?

Reply:
Yes i think so, but as i said, most of that is done in the frontend. I guess we have to ask Per about that.